inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/ClassRetransformationService.java

```diff
@@ -0,0 +1,145 @@
+    }
+
+    /**
+     * Reconfigures the jobs batch sizes and the inter batch delay on configuration changes.
+     *
+     * @param ev the change event
+     */
+    @EventListener
+    private void configEventListener(InspectitConfigChangedEvent ev) {
+        val settings = ev.getNewConfig().getInstrumentation().getInternal();
+        retransformJob.setBatchSizes(settings.getClassRetransformBatchSize());
+        retransformJob.setInterBatchDelay(settings.getClassRetransformInterBatchDelay());
+    }
+
+    /**
+     * This method is regularly called through {@link #retransformJob} to process elements from {@link #pendingClasses}.
+     * The goal is to get and remove batchSize classes from the queueand to invoke {@link Instrumentation#retransformClasses(Class[])} on them.
+     * <p>
+     * The package access is for testing.
+     */
+    void retransformClasses(int batchSize) {
+        val watch = new StopWatch();
+        if (!transformer.isShuttingDown()) {
+            Set<Class<?>> classesToRetransform = new HashSet<>();
+            try {
+                Optional<Class<?>> nextClass = CommonUtils.pollElementFromSet(pendingClasses);
+                while (classesToRetransform.size() < batchSize && nextClass.isPresent()) {
+                    Class<?> clazz = nextClass.get();
+                    pendingClasses.remove(clazz);
+
+                    //the last up-date check might have been a long time ago... is the retransformation still required?
+                    if (doesClassRequireRetransformation(clazz)) {
+                        classesToRetransform.add(clazz);
+                    }
+                    nextClass = CommonUtils.pollElementFromSet(pendingClasses);
+                }
+            } catch (Exception e) {
+                log.error("Error selecting Elements retransformation", e);
+            }
+
+            if (!classesToRetransform.isEmpty()) {
+                try {
+                    instrumentation.retransformClasses(classesToRetransform.toArray(new Class<?>[]{}));
+                    log.debug("Retransformed {} classes in {} ms, {} classes left to retransform", classesToRetransform.size(), watch.getElapsedMillis(), pendingClasses.size());
+                } catch (Exception e) {
+                    log.error("Error applying instrumentation:", e);
+                }
+            }
+        }
+    }
+
+    /**
+     * Todo: where to actually put this method... ? (Code duplication)
+     *
+     * @param clazz
+     * @return
+     */
+    private boolean doesClassRequireRetransformation(Class<?> clazz) {
+        val typeDescr = TypeDescription.ForLoadedType.of(clazz);
+        InstrumentationSettings instrSettings = env.getCurrentConfig().getInstrumentation();
+        val requestedConfig = ClassInstrumentationConfiguration.getFor(typeDescr, instrSettings, specialSensors);
+
+        val activeConfig = transformer.getActiveClassInstrumentationConfiguration(clazz);
```

Comment:
maybe the active configuration could be stored in some special service for this purpose, where the transformer would report on the instrumentation and others can ask for such things? We are not gaining much, so let's think..

Reply:
Moved it to the InstrumentationManager